File: sdks/python/examples/basic_usage.py
#!/usr/bin/env python3
"""Basic usage example for the Untrace Python SDK."""

import asyncio
import os
from untrace import UntraceClient


async def main():
    """Demonstrate basic SDK usage."""
    # Get API key from environment variable
    api_key = os.getenv("UNTRACE_API_KEY")
    if not api_key:
        print("Please set UNTRACE_API_KEY environment variable")
        return

    # Initialize the client
    async with UntraceClient(api_key=api_key) as client:
        # Example 1: Trace an LLM call
        print("Tracing LLM call...")
        trace = await client.trace(
            event_type="llm_call",
            data={
                "model": "gpt-4",
                "prompt": "What is the capital of France?",
                "response": "The capital of France is Paris.",
                "tokens_used": 15,
                "cost": 0.0003,
            },
            metadata={
                "user_id": "user123",
                "session_id": "session456",
                "request_id": "req789",
            }
        )
        print(f"✅ LLM call traced: {trace.id}")

        # Example 2: Trace a user action
        print("Tracing user action...")
        trace = await client.trace(
            event_type="user_action",
            data={
                "action": "button_click",
                "button_id": "submit_form",
                "page": "/dashboard",
            },
            metadata={
                "user_id": "user123",
                "session_id": "session456",
                "timestamp": "2023-01-01T12:00:00Z",
            }
        )
        print(f"✅ User action traced: {trace.id}")

        # Example 3: Retrieve a trace
        print(f"Retrieving trace: {trace.id}")
        retrieved_trace = await client.get_trace(trace.id)
        print(f"✅ Retrieved trace: {retrieved_trace.event_type}")


def sync_example():
    """Demonstrate synchronous SDK usage."""
    api_key = os.getenv("UNTRACE_API_KEY")
    if not api_key:
        print("Please set UNTRACE_API_KEY environment variable")
        return

    # Initialize the client
    client = UntraceClient(api_key=api_key)

    try:
        # Trace an event synchronously
        trace = client.trace_sync(
            event_type="api_call",
            data={
                "endpoint": "/api/v1/users",
                "method": "GET",
                "status_code": 200,
                "response_time_ms": 150,
            },
            metadata={
                "user_id": "user123",
                "ip_address": "192.168.1.1",
            }
        )
        print(f"✅ API call traced: {trace.id}")

    finally:
        # Always close the client
        client.close()


if __name__ == "__main__":
    print("=== Async Example ===")
    asyncio.run(main())

    print("\n=== Sync Example ===")
    sync_example()

File: sdks/elixir/examples/basic_usage.exs
#!/usr/bin/env elixir
# Basic usage example for the Untrace Elixir SDK

# Get API key from environment variable
api_key = System.get_env("UNTRACE_API_KEY")

if is_nil(api_key) do
  IO.puts("Please set UNTRACE_API_KEY environment variable")
  System.halt(1)
end

# Start the client
{:ok, client} = Untrace.Client.start_link(api_key: api_key)

IO.puts("=== Basic Usage Example ===")

# Example 1: Trace an LLM call
IO.puts("Tracing LLM call...")
{:ok, trace} = Untrace.Client.trace(client, %{
  event_type: "llm_call",
  data: %{
    model: "gpt-4",
    prompt: "What is the capital of France?",
    response: "The capital of France is Paris.",
    tokens_used: 15,
    cost: 0.0003
  },
  metadata: %{
    user_id: "user123",
    session_id: "session456",
    request_id: "req789"
  }
})
IO.puts("✅ LLM call traced: #{trace.id}")

# Example 2: Trace a user action
IO.puts("Tracing user action...")
{:ok, trace} = Untrace.Client.trace(client, %{
  event_type: "user_action",
  data: %{
    action: "button_click",
    button_id: "submit_form",
    page: "/dashboard"
  },
  metadata: %{
    user_id: "user123",
    session_id: "session456",
    timestamp: "2023-01-01T12:00:00Z"
  }
})
IO.puts("✅ User action traced: #{trace.id}")

# Example 3: Retrieve a trace
IO.puts("Retrieving trace: #{trace.id}")
{:ok, retrieved_trace} = Untrace.Client.get_trace(client, trace.id)
IO.puts("✅ Retrieved trace: #{retrieved_trace.event_type}")

# Example 4: Using helper functions
IO.puts("Using helper functions...")

# Create LLM trace using helper
llm_trace = Untrace.Types.create_llm_trace(
  "gpt-3.5-turbo",
  "Hello, world!",
  "Hello! How can I help you today?",
  10,
  5,
  %{user_id: "user123"}
)

{:ok, trace} = Untrace.Client.trace(client, llm_trace)
IO.puts("✅ LLM trace created with helper: #{trace.id}")

# Create user action trace using helper
user_trace = Untrace.Types.create_user_action_trace(
  "button_click",
  %{button_id: "submit", page: "/form"},
  %{user_id: "user123"}
)

{:ok, trace} = Untrace.Client.trace(client, user_trace)
IO.puts("✅ User action trace created with helper: #{trace.id}")

# Create API call trace using helper
api_trace = Untrace.Types.create_api_call_trace(
  "GET",
  "/api/v1/users",
  200,
  150,
  %{user_id: "user123"}
)

{:ok, trace} = Untrace.Client.trace(client, api_trace)
IO.puts("✅ API call trace created with helper: #{trace.id}")

# Stop the client
Untrace.Client.stop(client)
IO.puts("✅ Client stopped")
